beacon-chain/core/blocks/eth1_data_test.go

```diff
@@ -58,9 +58,10 @@ func TestEth1DataHasEnoughSupport(t *testing.T) {
 
 	for i, tt := range tests {
 		t.Run(fmt.Sprintf("%d", i), func(t *testing.T) {
-			c := params.BeaconConfig()
+			c := params.BeaconConfig().Copy()
```

Comment:
Where did this Copy method come from? Seems like a very easy way to misuse this API.

Reply:
I've added that method in #5691

I needed a method to copy state, so that we do not operate on the same pointer data. API, as used now, is: get pointer, update some value (changing the original object), override. I wanted to have a way to manipulate the params w/o side effects. So, I operate on copy of the object, when overriding I get a resetting function - so I can get back to the exact original state.

Thus, my understanding is that having a global state, which has its values mutated inside a unit test, is not a safe way to use the API. Hence, the proposed solution. Do you have any suggestions on how this should be done, instead?

Reply:
I have two ideas:
- Full copy on read
- Keep the original copy in memory, copy-on-write for reset.

I think that using Copy() will be difficult to enforce unless we add another static analysis check and that it won't be very clear to users that they should use the copy method in tests. I wish we had immutable objects in go :(

Reply:
I suppose that full copy on read (in `params.BeaconChain()`, `params.MainnetConfig()`, and `params.MinimalSpecConfig()`) is the most straight-forward way to proceed.

So, I will make sure that `params.BeaconConfig()` returns copy of the object, and have the sth like this:
```go
// Updating some params
config := params.BeaconConfig()
config.foo = "bar" // safe to do on copy
resetCfg := params.OverrideBeaconConfigWithReset(config)
defer resetCfg()

// Using certain preset config
resetCfg := params.OverrideBeaconConfigWithReset(params.MainnetConfig())
defer resetCfg()
```

I will also reduce scope of `params.copy()` to package-level.
Thoughts?

Reply:
The side effect of copy on read is that there will be much more memory churn in the production code path. Maybe we can try the other idea. Something like

originalDefaultConfig := deepCopy(defaultConfig)

func reset() {
  defaultConfig = deepCopy(originalDefaultConfig)
}

Reply:
The side effect of copy on read is that there will be much more memory churn in the production code path. Maybe we can try the other idea. Something like

originalDefaultConfig := deepCopy(defaultConfig)

func reset() {
  defaultConfig = deepCopy(originalDefaultConfig)
}

Reply:
So, keeping in mind that we want to:
- continue returning pointers on read, in order not to introduce more memory churn
- be able to modify config params individually and as a whole (that's changing mainnet/miniconf)
- make sure that once test case is done, teardown code is invoked, which restores previous global state - i.e. makes sure that there's no side-effects leaked out of test case

I've updated code to use testing.T.Cleanup, which is available as of Go 1.14, and allows to register cleanup function with the test case. So, whenever, within a testcase params modification are needed, one needs to invoke `params.SetupTestConfigCleanup(t *testing.T)`, which does the following:
```golang
// SetupTestConfigCleanup preserves configurations allowing to modify them 
// within tests without any restrictions, everything is restored after the test.
func SetupTestConfigCleanup(t *testing.T) {
	prevDefaultBeaconConfig := defaultBeaconConfig.Copy()
	prevBeaconConfig := beaconConfig.Copy()
	t.Cleanup(func() {
		defaultBeaconConfig = prevDefaultBeaconConfig
		beaconConfig = prevBeaconConfig
	})
}
```

Now, within individual tests not much is changing, for example:
```golang
func TestDiscv5_AddRetrieveForkEntryENR(t *testing.T) {
	params.SetupTestConfigCleanup(t) // Configure teardown function
	c := params.BeaconConfig() // Get pointer and modify values as needed
	c.ForkVersionSchedule = map[uint64][]byte{
		0: params.BeaconConfig().GenesisForkVersion,
		1: {0, 0, 0, 1},
	}
	nextForkEpoch := uint64(1)
	nextForkVersion := []byte{0, 0, 0, 1}
	c.NextForkEpoch = nextForkEpoch
	c.NextForkVersion = nextForkVersion
	params.OverrideBeaconConfig(c)
        // ...
        // Once test is finished, cleanup code is called.
}
```